compiler/rustc_typeck/src/check/wfcheck.rs

```diff
@@ -258,251 +258,168 @@ pub fn check_trait_item(tcx: TyCtxt<'_>, def_id: LocalDefId) {
-        // relationship to the other region arguments. If there is an
-        // outlives relationship, then we want to ensure that is
-        // reflected in a where clause on the GAT itself.
-        for (region_a, region_a_idx) in &regions {
-            // Ignore `'static` lifetimes for the purpose of this lint: it's
-            // because we know it outlives everything and so doesn't give meaninful
-            // clues
-            if let ty::ReStatic = region_a {
-                continue;
-            }
-            for (region_b, region_b_idx) in &regions {
-                if region_a == region_b {
+            let mut new_required_bounds: Option<FxHashSet<ty::Predicate<'_>>> = None;
+            for item in assoc_items {
+                let item_def_id = item.id.def_id;
+                // Skip our own GAT, since it would blow away the required bounds
+                if item_def_id == gat_def_id {
                     continue;
                 }
-                if let ty::ReStatic = region_b {
+
+                let item_hir_id = item.id.hir_id();
+                let param_env = tcx.param_env(item_def_id);
+
+                let Some(item_required_bounds) = (match item.kind {
+                    hir::AssocItemKind::Fn { .. } => {
+                        let sig: ty::FnSig<'_> = tcx.liberate_late_bound_regions(
+                            item_def_id.to_def_id(),
+                            tcx.fn_sig(item_def_id),
+                        );
+                        gather_gat_bounds(
+                            tcx,
+                            param_env,
+                            item_hir_id,
+                            sig.output(),
+                            &sig.inputs().iter().copied().collect(),
+                            gat_def_id,
+                            gat_generics,
+                        )
+                    }
+                    hir::AssocItemKind::Type => {
+                        // If our associated item is a GAT with missing bounds, add them to
+                        // the param-env here. This allows this GAT to propagate missing bounds
+                        // to other GATs.
+                        let param_env = augment_param_env(
+                            tcx,
+                            item_def_id.to_def_id(),
+                            param_env,
+                            required_bounds_by_item.get(&item_def_id),
+                        );
+                        let bounds: Vec<_> = tcx
+                            .explicit_item_bounds(item_def_id)
+                            .iter()
+                            .map(|(predicate, _)| {
+                                tcx.liberate_late_bound_regions(
+                                    item_def_id.to_def_id(),
+                                    predicate.kind(),
+                                )
+                            })
+                            .collect();
```

Comment:
Again comes down to perf, but might be interesting to think how we could avoid collecting into a `Vec` here

Reply:
`Iterator`s aren't `TypeFoldable` are they? But also maybe we can just pass the original bounds as a `&'tcx List<_>` and then do the late bound liberation in the `GatSubstsCollector` below.

Actually... we already do liberate binders in `GatSubstsCollector`, which I added in #93227 a while back. Okay then. Let me just not liberate in this method at all and leave it to that.

That being said, do you know what the significance of the def-id of the first argument to `liberate_late_bound_regions` is? Maybe #93227 is wrong by passing in the GAT def_id there, and we want to pass the item's def_id instead.

Reply:
It's the "scope" - don't think it really matters here.